Trustroots/trustroots
# Show API documentation at /api/

#### Proposed Changes

- auto-generate API documentation in OpenAPI format with `npm run generate-api-docs` (the documentation isn't perfect, but it's good enough, much better than doing it manually) using swagger-autogen
- display API documentation with Swagger-UI on `/api/` path
- hide the API documentation behind feature flag `{ featureFlags: { apidocs: true } }` in `./config/env/local.js`
  - by default it shows in development, but not in production
- generate up-to-date documentation before every commit with husky
- additional fix: update leaflet-active-area library to latest version from npm; otherwise webhooks fail to install node_modules when they try to fetch the library from github, because of some weird ssh authentication reason.

The OpenAPI json, and files for Swagger-UI are stored in /apidocs/ folder. It can be moved elsewhere if somebody prefers

#### Motivation

For people who want to develop e.g. mobile app on top of Trustroots API, this can be useful resource

#### Testing instructions

1. Update node_modules by running `npm ci`

1. Try to delete `/apidocs/openapi.json`, then run `npm run generate-api-docs`, and it should re-appear

1. Run the app locally, and go to http://localhost:3000/api/, and you should see
  ![image](https://github.com/Trustroots/trustroots/assets/7449720/288cb171-c623-4b21-9ac9-be4bcda04539)

1. If running the whole app locally is too much for you, you can
  - install e.g. http-server: `sudo npm install -g http-server`
  - in terminal, go to Trustroots repository that you cloned locally
  - go to this branch `git switch api-docs`
  - `cd apidocs/`
  - run the server with command `http-server`
  - in your browser, go to http://localhost:8080, and see the documentation
